File: ts/src/tree.ts
export class BinarySearchTree {
  private root: Node | null = null;

  insert(value: number): void {
    this.root = this.insertNode(this.root, value);
  }

  inorder(): number[] {
    const res: number[] = [];
    const dfs = (node: Node | null) => {
      if (!node) return;
      dfs(node.left);
      res.push(node.value);
      dfs(node.right);
    };
    dfs(this.root);
    return res;
  }

  private insertNode(node: Node | null, value: number): Node {
    if (!node) return new Node(value);
    if (value < node.value) node.left = this.insertNode(node.left, value);
    else node.right = this.insertNode(node.right, value);
    return node;
  }
}

class Node {
  value: number;
  left: Node | null = null;
  right: Node | null = null;

  constructor(value: number) {
    this.value = value;
  }
}

export class MinHeap {
  private data: number[] = [];

  push(v: number): void {
    this.data.push(v);
    this.siftUp(this.data.length - 1);
  }

  pop(): number {
    if (this.data.length === 0) throw new Error("heap is empty");
    const top = this.data[0];
    const last = this.data.pop()!;
    if (this.data.length > 0) {
      this.data[0] = last;
      this.siftDown(0);
    }
    return top;
  }

  private siftUp(i: number) {
    while (i > 0) {
      const p = (i - 1) >> 1;
      if (this.data[p] <= this.data[i]) break;
      [this.data[p], this.data[i]] = [this.data[i], this.data[p]];
      i = p;
    }
  }

  private siftDown(i: number) {
    const n = this.data.length;
    while (true) {
      let s = i;
      const l = i * 2 + 1;
      const r = i * 2 + 2;
      if (l < n && this.data[l] < this.data[s]) s = l;
      if (r < n && this.data[r] < this.data[s]) s = r;
      if (s === i) break;
      [this.data[s], this.data[i]] = [this.data[i], this.data[s]];
      i = s;
    }
  }
}

File: rust/src/tree.rs
pub struct BstNode {
    pub val: i32,
    pub left: Option<Box<BstNode>>,
    pub right: Option<Box<BstNode>>,
}

impl BstNode {
    pub fn new(val: i32) -> Self {
        Self {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn bst_insert(root: &mut Option<Box<BstNode>>, val: i32) {
    match root {
        None => *root = Some(Box::new(BstNode::new(val))),
        Some(node) => {
            if val < node.val {
                bst_insert(&mut node.left, val);
            } else {
                bst_insert(&mut node.right, val);
            }
        }
    }
}

pub fn inorder(root: &Option<Box<BstNode>>, out: &mut Vec<i32>) {
    if let Some(node) = root {
        inorder(&node.left, out);
        out.push(node.val);
        inorder(&node.right, out);
    }
}

pub struct MinHeap {
    data: Vec<i32>,
}

impl MinHeap {
    pub fn new() -> Self {
        Self { data: Vec::new() }
    }

    pub fn push(&mut self, x: i32) {
        self.data.push(x);
        let mut i = self.data.len() - 1;
        while i > 0 {
            let p = (i - 1) / 2;
            if self.data[p] <= self.data[i] {
                break;
            }
            self.data.swap(i, p);
            i = p;
        }
    }

    pub fn pop(&mut self) -> Option<i32> {
        if self.data.is_empty() {
            return None;
        }
        let top = self.data[0];
        let last = self.data.pop().unwrap();
        if !self.data.is_empty() {
            self.data[0] = last;
            self.sift_down(0);
        }
        Some(top)
    }

    fn sift_down(&mut self, mut i: usize) {
        let n = self.data.len();
        loop {
            let mut s = i;
            let l = i * 2 + 1;
            let r = i * 2 + 2;
            if l < n && self.data[l] < self.data[s] {
                s = l;
            }
            if r < n && self.data[r] < self.data[s] {
                s = r;
            }
            if s == i {
                break;
            }
            self.data.swap(i, s);
            i = s;
        }
    }
}
